// @buffetjs/custom/build/esm/components/List/index.js
/**
 *
 * List
 *
 */
import React from 'react';
import PropTypes from 'prop-types';
import { List as ListCompo, ListHeader, Padded } from '@buffetjs/core';
import { Card } from '@buffetjs/styles';

function List(_ref) {
  var title = _ref.title,
      subtitle = _ref.subtitle,
      button = _ref.button,
      isLoading = _ref.isLoading,
      items = _ref.items,
      customRowComponent = _ref.customRowComponent;
  return /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(Padded, {
    right: true,
    left: true,
    size: "md"
  }, /*#__PURE__*/React.createElement(ListHeader, {
    title: title,
    subtitle: subtitle,
    button: button
  })), /*#__PURE__*/React.createElement(ListCompo, {
    items: items,
    isLoading: isLoading,
    customRowComponent: customRowComponent
  }));
}

List.defaultProps = {
  button: null,
  customRowComponent: null,
  isLoading: false,
  items: [],
  title: null,
  subtitle: null
};
List.propTypes = {
  button: PropTypes.shape({
    color: PropTypes.string,
    icon: PropTypes.bool,
    type: PropTypes.string
  }),
  customRowComponent: PropTypes.func,
  isLoading: PropTypes.bool,
  items: PropTypes.instanceOf(Array),
  subtitle: PropTypes.string,
  title: PropTypes.string
};
export default List;
